opm/common/utility/numeric/VectorUtil.hpp

```diff
@@ -0,0 +1,94 @@
+#ifndef VECTORUTIL_H
+#define VECTORUTIL_H
+
+#include <tuple>
+#include <vector>
+#include <functional>
+#include <stdexcept> 
+#include <type_traits>
+
+namespace VectorUtil {
+
+// Implementation of generation General Operation between two vectors of the same type
+template <typename T, typename Operation>
+std::vector<T> vectorOperation(const std::vector<T>& vecA, const std::vector<T>& vecB, Operation op) {
+    if (vecA.size() != vecB.size()) {
+        throw std::invalid_argument("Error: Vectors must have the same size!"); // Throwing exception
+    }
+    std::vector<T> result;
+    result.reserve(vecA.size());
+    // Use std::transform with the passed operation
+    std::transform(vecA.begin(), vecA.end(), vecB.begin(), std::back_inserter(result), op);
+    return result;
+}
+
+
+// A simple utility function to calculate area of a rectangle
+std::tuple<std::array<double,4>, std::array<double,4>, std::array<double,4>> 
+appendNode(const std::array<double,3>&, const std::array<double,3>&, 
+           const std::array<double,3>&, const double&, const double&, 
+           const double&);
+template <typename T>
+std::vector<T> filterArray(const std::vector<std::size_t>& X, const std::vector<int>& ind){
+    std::vector<T> filtered_vectorX(ind.size(),0);
+    for (std::size_t index = 0; index < ind.size(); index++) {
+        filtered_vectorX[index] = X[ind[index]];
+    }
+    return filtered_vectorX;
+};       
+
+// T type of the object
+// Rout type of the object output 
+// Method type of method
+template <typename T,  typename Rout, typename Rin = std::size_t, typename Method, typename... Args>
```

Comment:
Slight simplification: If you order the template parameters as
```C++
template <typename Rout, typename T, typename Rin, typename Method, typename... Args>
```
then the call site need only specify the `Rout` as an explicit template argument.  The rest will be deduced from the function arguments.  In other words, instead of
```C++
auto result = callMethodForEachInputOnObject<T, Rout, Rin, Method, Args...>(...);
```
you can get by with the simpler
```C++
auto result = callMethodForEachInputOnObject<Rout>(...);
```
instead.  For what it's worth, this is what `std::make_shared<>()` and `std::make_unique<>()` do.  They need the type to create as an explicit template argument, but the rest of the template parameters are deduced from the function arguments.

Reply:
Let's us discuss a bit about some of this modification today. I have some questions.